### PR TITLE
Implement CONNECT tunnels for HTTP/2 prior knowledge

### PR DESCRIPTION
**Pull request by swankjesse**
Closes: https://github.com/square/okhttp/issues/7289

### CODE REVIEW

File: okhttp/src/jvmMain/kotlin/okhttp3/Route.kt [hunk 1]
```diff
@@ -64,12 +64,16 @@ class Route(
...
    * See [RFC 2817, Section 5.2][rfc_2817].
    *
    * [rfc_2817]: http://www.ietf.org/rfc/rfc2817.txt
    */
-  fun requiresTunnel(): Boolean = address.sslSocketFactory != null && proxy.type() == Proxy.Type.HTTP
+  fun requiresTunnel(): Boolean {
+    if (proxy.type() != Proxy.Type.HTTP) return false
+    return (address.sslSocketFactory != null) ||
+      (Protocol.H2_PRIOR_KNOWLEDGE in address.protocols)
+  }
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/Route.kt:76]: I’m delighted at how easy this turned out! The rest is tests!

**Review by juliaogris — APPROVED** [review]

File: okhttp/src/jvmTest/java/okhttp3/internal/http2/HttpOverHttp2Test.kt [hunk 2]
```diff
@@ -2045,27 +2041,77 @@ class HttpOverHttp2Test {
...
 
+  /** Respond to a proxy authorization challenge.  */
+  @ParameterizedTest
+  @ArgumentsSource(ProtocolParamProvider::class)
+  fun proxyAuthenticateOnConnect(protocol: Protocol, mockWebServer: MockWebServer) {
+    setUp(protocol, mockWebServer)
+    server.enqueue(
+      MockResponse()
+        .inTunnel()
+        .setResponseCode(407)
```
Comment (yschimke) [inline okhttp/src/jvmTest/java/okhttp3/internal/http2/HttpOverHttp2Test.kt:2075]: This answers my question on other PR.

File: okhttp/src/jvmTest/java/okhttp3/internal/http2/HttpOverHttp2Test.kt [hunk 3]
```diff
@@ -2045,27 +2041,77 @@ class HttpOverHttp2Test {
       .proxy(server.toProxyAddress())
       .build()
 
-    val call = client.newCall(
-      Request(
-        server.url("/").resolve("//android.com/foo")!!
-      )
-    )
+    val url = server.url("/").resolve("//android.com/foo")!!
```
Comment (yschimke) [inline okhttp/src/jvmTest/java/okhttp3/internal/http2/HttpOverHttp2Test.kt:2044]: Existing behaviour, but where does this format get handled?

**Review by yschimke — APPROVED** [review]
Nice

File: okhttp/src/jvmTest/java/okhttp3/internal/http2/HttpOverHttp2Test.kt (hunk 3 above)
Comment (swankjesse) [reply okhttp/src/jvmTest/java/okhttp3/internal/http2/HttpOverHttp2Test.kt:2044]: It’s in HttpUrl. Two leading slashes keeps the scheme only!

File: okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/ConnectPlan.kt [hunk 4]
```diff
@@ -160,6 +160,14 @@ class ConnectPlan(
       }
 
       if (route.address.sslSocketFactory != null) {
+        // Assume the server won't send a TLS ServerHello until we send a TLS ClientHello. If
+        // that happens, then we will have buffered bytes that are needed by the SSLSocket!
+        // This check is imperfect: it doesn't tell us whether a handshake will succeed, just
+        // that it will almost certainly fail because the proxy has sent unexpected data.
+        if (source?.buffer?.exhausted() == false || sink?.buffer?.exhausted() == false) {
+          throw IOException("TLS tunnel buffered too many bytes!")
+        }
```
Comment (swankjesse) [inline okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/ConnectPlan.kt:169]: FYI @yschimke I had to move this check to get tests to stop flaking. In TLS handshakes the server doesn’t send until it receives data from the client. In HTTP/2 the server and client can both send immediately!

**Pull request merged**